pyroomacoustics/libroom_src/room.cpp

```diff
@@ -276,3 +276,422 @@ int Room::image_source_shoebox(
+	Eigen::Vector3f min_point(minX, minY, minZ);
+	
+	return (min_point - max_point).norm() + 1;
+
+}
+
+
+
+Eigen::VectorXf Room::next_wall_hit(
+					const Eigen::VectorXf &start,
+					const Eigen::VectorXf &end,
+					Eigen::Ref<Vector1i> next_wall_index){
+						
+	/* Computes the next next wall_hit position given a segment defined
+	 * by its endpoints. This method also stores the index of the wall
+	 * that contains this next hit point (we needed to pass it inside a 
+	 * Vector1i to be able to call this function from python).
+	 * 
+	 * If no wall is intersected, then the next_wall_index will be -1*/
+						
+	Eigen::VectorXf result;
+	result.resize(dim);
+	
+	next_wall_index[0] = -1;
+	
+	// Upperbound on the min distance that we could find
+	float min_dist(max_dist);
+	
+	for (size_t i(0); i < walls.size(); ++i){
+		
+		Wall w = walls[i];
+		
+		// To store the result of this iteration
+		Eigen::VectorXf temp_hit;
+		temp_hit.resize(dim);
+
+
+		// As a side effect, temp_hit gets a value (VectorXf) here
+		bool intersects = (w.intersection(start, end, temp_hit) > -1);
+		
+		//~ std::cout << "\n\n\ntemp_hit : " << temp_hit[0]<< " " << temp_hit[1]<< " " << temp_hit[2] << std::endl;
+		//~ std::cout << "start : " << start[0]<< " " << start[1]<< " " << start[2] << std::endl;
+		//~ std::cout << "end : " << end[0]<< " " << end[1]<< " " << end[2] << std::endl;
+
+		if (intersects){
+				
+			
+			
+			float temp_dist = (start-temp_hit).norm();
+			
+			if (temp_dist > libroom_eps and temp_dist < min_dist){
+				
+				min_dist = temp_dist;
+				result = temp_hit;
+				
+				// update the index of the wall that is hit
+				next_wall_index[0] = i;
+			}
+		}
+		
+	}
+	return result;
+}					
+					
+					
+bool Room::scat_ray(const Wall &last_wall,
+					const Eigen::VectorXf &hit_point,
+					float radius,
+					float scat_energy,
+					float travel_time,
+					float time_thres,
+					float sound_speed,
+					std::vector<entry> &output){
+						
+	/*
+	 * Trace a one-hop scattering ray from the last wall hit to the microphone.
+	 * In case the scattering ray can reach the microphone (no wall in
+	 * between), then we add an entry to the output passed by reference.
+	 * 
+    - last_wall: The wall object where last_hit is located
+    - radius : The radius of the circle/sphere microphone
+    - last_hit: An array of length 2 or 3 defining the last wall hit position
+    - scat_energy: The energy of the scattering ray
+    - actual_travel_time: The cumulated travel time of the ray
+    - time_thresh: The time threshold of the ray
+    - sound_speed: The speed of sound
+    * 
+    * Return : true if the scattered ray reached the mic, false otw
+	*/
+	
+	bool result = false;
+	
+	Vector1i next_wall_idx(-1);
+	
+	next_wall_hit(hit_point, mic_pos, next_wall_idx);
+	
+	
+	// No wall intersecting the direct scattered ray
+	if (next_wall_idx[0] == -1){
+		
+		VectorXf mic_hit = mic_intersection(hit_point, mic_pos, mic_pos, radius);
```

Comment:
I think we are taking quite a performance hit from computing the intersection with the receiver surface. This might not be necessary actually. Just using the distance to the closest point to the receiver on the straight line should be good enough.

Reply:
What do you mean by "_using the distance to the closest point to the receiver on the straight line_" ?

I could indeed trace a segment [wall_hit, mic_center] and then take the point that is  'mic_radius' away from the mic_center.

Was that what you meant ?

Reply:
No need to do that. I'm thinking that we can use directly the same computations that check for the intersection. This is related to the above comment that describes how to compute the distance from line to point.

1) Compute the distance from ray (line) to microphone (point)
2) If the distance is less than the receiver radius, there is an intersection
3) Use the point closest to the microphone on the ray to compute the distance. In fact, the traveled distance is already computed as a by-product of the intersection check.

I know that this makes the distance 'inexact', but if you think about it, the ratio of the error to the traveled distance is in fact quite small in most cases. This should be good enough for us. We need to keep in mind that the ray tracing process itself is an approximation of the true propagation of sound in space.

Reply:
I see your point and in fact this is already what I do in the intersects_mic function.

But the 'distance' variable should store the distance between the last wall_hit and the precise mic_hit. 

So for me, if we want to use an approximation, it would make more sense to  simply take the distance between the last wall_hit and the microphone center.

(Also for the precise code that you quoted, the scattered ray is shooted directly towards the microphone center. So, in that case I can obviously simplify my formula for the distance and juste obtain it like this : `distance = (hit_point - mic_pos).norm() - radius`)

Reply:
Right, then let's do that for scattered rays.

For the other rays, here again we are not trying to find the exact timing, but more the overall energy that hits the receiver in some space/time slot. A different of a few tens of centimeters will not change the overall result.